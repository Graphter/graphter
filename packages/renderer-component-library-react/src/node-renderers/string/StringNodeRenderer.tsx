import React, { useState } from "react";
import s from './StringNodeRenderer.pcss'
import { NodeRendererProps } from "microo-core";
import { createDefault } from "malle-renderer-react";

function StringNodeRenderer(
  {
    config,
    ancestryConfig,
    jsonPointer,
    nodeData,
    setNodeDataValue,
    validationResults,
    DataProvider,
  }: NodeRendererProps
){
  const isNew = typeof nodeData === 'undefined'
  if(isNew) nodeData = createDefault(config, '')
  const [ touched, setTouched ] = useState(false);
  return (
    <>
      <input
        type='text'
        id={config.id}
        value={nodeData}
        data-testid='string-input'
        onChange={(e) => {
          if(!touched) setTouched(true)
          setNodeDataValue && setNodeDataValue(e.currentTarget.value);
        }} />
      {touched && validationResults && validationResults.map((result, i) => (
        <div className={s.error} key={i}>{result.errorMessage}</div>
      ))}
    </>
  )
}
StringNodeRenderer.type = 'string'
export default StringNodeRenderer