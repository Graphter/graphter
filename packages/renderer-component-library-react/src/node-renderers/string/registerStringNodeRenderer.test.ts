import StringNodeRenderer, { registerStringNodeRenderer } from "../string";

describe('registerStringNodeRenderer()', () => {
  it('should return the correct registration', async () => {
    const result = registerStringNodeRenderer()
    expect(result.type).toBe('string')
    expect(result.getChildData).toBeUndefined()
    expect(result.getChildPaths).toBeUndefined()
    expect(result.Renderer).toBe(StringNodeRenderer)
    expect(result.options).toBeUndefined()
  })
  it('should override the type if supplied in options', async () => {
    const result = registerStringNodeRenderer({ type: 'custom-string' })
    expect(result.type).toBe('custom-string')
  })
})