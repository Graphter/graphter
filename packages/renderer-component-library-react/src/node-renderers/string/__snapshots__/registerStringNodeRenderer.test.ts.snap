// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerStringNodeRenderer() should return the correct registration 1`] = `
Object {
  "Renderer": [Function],
  "description": "Manage small amounts of plain text",
  "name": "Text",
  "options": undefined,
  "type": "string",
}
`;
